src/sentry/workflow_engine/processors/workflow_preview.py

```diff
@@ -0,0 +1,140 @@
+
+
+def preview_conditions(
+    logic_type: DataConditionGroup.Type, data_conditions: list[DataCondition], group_ids: set[int]
+) -> tuple[set[int], set[int]]:
+    # groups that would have triggered each condition
+    conditions_to_triggered_groups = {}
+
+    for condition in data_conditions:
+        conditions_to_triggered_groups[condition.id] = condition.get_preview_groups(group_ids)
+
+    groups_to_check = set(group_ids)
+    groups_meeting_conditions: set[int] = set()
+
+    triggered_groups: list[set[int]] = list(conditions_to_triggered_groups.values())
+
+    match logic_type:
+        case DataConditionGroup.Type.ALL:
+            # ALL logic type: intersection of all the groups that meet any condition (= met all conditions)
+            # should only continue checking groups that have already met these conditions
+            groups_meeting_conditions = set.intersection(*triggered_groups)
+            groups_to_check = groups_meeting_conditions
+        case DataConditionGroup.Type.ANY:
+            # ANY logic type: union of all the groups that meet any condition
+            # should only continue checking groups that have not already met these conditions
+            groups_meeting_conditions = set().union(*triggered_groups)
+            groups_to_check -= groups_meeting_conditions
+        case DataConditionGroup.Type.ANY_SHORT_CIRCUIT:
+            groups_meeting_conditions = set().union(*triggered_groups)
+            groups_to_check -= groups_meeting_conditions
+        case DataConditionGroup.Type.NONE:
+            # NONE logic type: all the group ids - intersection of all the groups that meet any condition (= did not meet any condition)
+            # should only continue checking groups that did not meet any condition
+            groups_meeting_conditions = groups_to_check - set().union(*triggered_groups)
+            groups_to_check = groups_meeting_conditions
+
+    return groups_to_check, groups_meeting_conditions
+
+
+def preview_data_condition_group(data_condition_group_id: int, group_ids: set[int]) -> set[int]:
+    try:
+        dcg = DataConditionGroup.objects.get_from_cache(id=data_condition_group_id)
+    except DataConditionGroup.DoesNotExist:
+        logger.exception(
+            "DataConditionGroup does not exist",
+            extra={"id": data_condition_group_id},
+        )
+        return set()
+
+    try:
+        logic_type = DataConditionGroup.Type(dcg.logic_type)
+    except ValueError:
+        logger.exception(
+            "Invalid DataConditionGroup.logic_type found in process_data_condition_group",
+            extra={"logic_type": dcg.logic_type},
+        )
+        return set()
+
+    conditions = get_data_conditions_for_group(data_condition_group_id)
+
+    # check fast conditions first
+    fast_conditions, slow_conditions = split_conditions_by_speed(conditions)
+
+    # TODO: early return if they are invalid conditon pairs? see src/sentry/rules/history/preview.py VALID_CONDITION_PAIRS
```

Comment:
https://github.com/getsentry/sentry/blob/6a4af04cdf7d3053fa6d2c70550334b543984556/src/sentry/rules/history/preview.py#L53-L59

Reply:
a lot of other early return conditions in the existing implementation
https://github.com/getsentry/sentry/blob/6a4af04cdf7d3053fa6d2c70550334b543984556/src/sentry/rules/history/preview.py#L74-L95